contracts/LidoMatic.sol

```diff
@@ -0,0 +1,476 @@
+     */
+    function requestWithdraw(uint256 _amount) external notPaused {
+        // Add a function that converts MATIC to StMATIC and reverse
+        // Check how many StMATIC does a caller have
+        // Add a mapping between a validator address and a nonce (starts at 1)
+        // Increment nonce by 1 everytime a user calls a withdrawal function
+        // Burn StMATIC after checking if _amount satisfies user's balance
+        // A nonce is dependent on a validator
+        // Add a nonce per user
+        Operator.OperatorShare[] memory operatorShares = nodeOperator
+            .getOperatorShares();
+
+        if (lastWithdrawnValidatorId > operatorShares.length - 1) {
+            lastWithdrawnValidatorId = 0;
+        }
+
+        address validatorShare = operatorShares[lastWithdrawnValidatorId]
+            .validatorShare;
+
+        uint256 callerBalance = balanceOf(msg.sender);
+
+        require(callerBalance >= _amount, "Invalid amount");
+
+        _burn(msg.sender, _amount);
+
+        uint256 amountInMATIC = getUserBalanceInMATIC();
+
+        sellVoucher_new(validatorShare, amountInMATIC, type(uint256).max);
+
+        if (validator2Nonce[validatorShare] == 0) {
+            validator2Nonce[validatorShare] = 1;
+        } else {
+            validator2Nonce[validatorShare]++;
+        }
+
+        if (user2Nonce[msg.sender] == 0) user2Nonce[msg.sender] = 1;
+
+        user2Nonce[msg.sender]++;
+
+        RequestWithdraw[] storage requestWithdraws = user2WithdrawRequest[
+            msg.sender
+        ];
+
+        requestWithdraws.push(
+            RequestWithdraw(
+                amountInMATIC,
+                validator2Nonce[validatorShare],
+                block.timestamp,
+                validatorShare,
+                true
+            )
+        );
+
+        lastWithdrawnValidatorId++;
+    }
+
+    /**
+     * @dev Delegates tokens to validator share contract
+     */
+    function delegate() external auth(DAO) {
+        Operator.OperatorShare[] memory operatorShares = nodeOperator
+            .getOperatorShares();
+
+        uint256 amountPerValidator = totalBuffered / operatorShares.length;
+        uint256 remainder = totalBuffered % operatorShares.length;
+
+        for (uint256 i = 0; i < operatorShares.length; i++) {
+            IERC20(token).approve(
+                operatorShares[i].validatorShare,
+                amountPerValidator
+            );
+
+            buyVoucher(operatorShares[i].validatorShare, amountPerValidator, 0);
+
+            validator2DelegatedAmount[
+                operatorShares[i].validatorShare
+            ] = amountPerValidator;
+        }
+
+        totalDelegated += totalBuffered - remainder;
+        totalBuffered = remainder;
+    }
+
+    /**
+     * @dev Claims tokens from validator share and sends them to the
+     * user if his request is in the userToWithdrawRequest
+     */
+    function claimTokens() external {
```

Comment:
I know that was my idea to do it like this :) but I just realized one thing.

The actual loop only processes the last request. Let's say the user requested withdraw 10 days ago and then request another withdraw today so even if the first request will be ready in let's say 5 days he can not withdraw it because the second one will be ready in 10days.

The user2WithdrawRequest is public and we can access it from the Front-end.
If we add the requestId is arg for claimTokens we can avoid using a loop and the user can call any request.

Reply:
This loop actually processes first request, not the last one. So if the user has 2 requests pending, he will be able to claim the newest one only after claiming the oldest one.